File: src/test/java/com/thebluehats/server/game/utils/EnchantLoreParserTest.java
package com.thebluehats.server.game.utils;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;

import org.bukkit.ChatColor;
import org.junit.Test;

public class EnchantLoreParserTest {
    @Test
    public void InsertsProperVariablesBasedOnEnchantLevel() {
        EnchantLoreParser enchantLoreParser = new EnchantLoreParser("A {0} {1} {2} C");

        String[][] variables = new String[3][];
        variables[0] = new String[] { "1", "2", "3" };
        variables[1] = new String[] { "A", "B", "C" };
        variables[2] = new String[] { "D", "E", "F" };

        enchantLoreParser.setVariables(variables);

        ArrayList<String> expectedLore = new ArrayList<>();
        expectedLore.add(ChatColor.GRAY + "A 2 B E C");

        ArrayList<String> parsedLore = enchantLoreParser.parseForLevel(2);

        assertEquals(expectedLore, parsedLore);
    }

    @Test
    public void CorrectlyInsertsSingleVariable() {
        EnchantLoreParser enchantLoreParser = new EnchantLoreParser("{0}");

        enchantLoreParser.setSingleVariable("A", "B", "C");

        ArrayList<String> expectedLore = new ArrayList<>();
        expectedLore.add(ChatColor.GRAY + "A");

        ArrayList<String> parsedLore = enchantLoreParser.parseForLevel(1);

        assertEquals(expectedLore, parsedLore);
    }

    @Test
    public void AppendsLore() {
        int level = 1;

        EnchantLoreParser enchantLoreParser = new EnchantLoreParser("A B C");
        enchantLoreParser.addTextIf(level == 1, "<br/>A");

        ArrayList<String> expectedLore = new ArrayList<>();
        expectedLore.add(ChatColor.GRAY + "A B C");
        expectedLore.add(ChatColor.GRAY + "A");

        ArrayList<String> parsedLore = enchantLoreParser.parseForLevel(level);

        assertEquals(expectedLore, parsedLore);
    }
}

File: src/test/kotlin/com/thebluehats/server/game/utils/EnchantLoreParserTest.kt
package com.thebluehats.server.game.utils

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe

class EnchantLoreParserTest : DescribeSpec({
    describe("enchant lore parser") {
        it("replaces tags with the proper chat color and linebreak") {
            val elp = EnchantLoreParser("<red>red</red><br/><light-purple>purple</light-purple>")

            val expectedLore = arrayListOf(
                "§7§cred§7",
                "§7§dpurple§7"
            )

            elp.parseForLevel(1) shouldBe expectedLore
        }

        it("replaces a single variable") {
            val elp = EnchantLoreParser("Hello world {0}!")

            val expectedL1Lore = arrayListOf("§7Hello world 1!")
            val expectedL2Lore = arrayListOf("§7Hello world 2!")
            val expectedL3Lore = arrayListOf("§7Hello world 3!")

            elp.setSingleVariable("1", "2", "3")

            elp.parseForLevel(1) shouldBe expectedL1Lore
            elp.parseForLevel(2) shouldBe expectedL2Lore
            elp.parseForLevel(3) shouldBe expectedL3Lore
        }

        it("replaces multiple variables") {
            val elp = EnchantLoreParser("Hello world {0} {1}!")

            val vars: Array<Array<String>?> = arrayOf(
                arrayOf("a", "b", "c"),
                arrayOf("x", "y", "z")
            )

            val expectedL1Lore = arrayListOf("§7Hello world a x!")
            val expectedL2Lore = arrayListOf("§7Hello world b y!")
            val expectedL3Lore = arrayListOf("§7Hello world c z!")

            elp.setVariables(vars)

            elp.parseForLevel(1) shouldBe expectedL1Lore
            elp.parseForLevel(2) shouldBe expectedL2Lore
            elp.parseForLevel(3) shouldBe expectedL3Lore
        }
    }
})
